File: platforms/documentation/docs/src/snippets/reference/plugin-development/readOnlyManagedProperty/groovy/buildSrc/src/main/java/Download.java
import org.gradle.api.DefaultTask;
import org.gradle.api.provider.Property;
import org.gradle.api.provider.Provider;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.Internal;
import org.gradle.api.tasks.TaskAction;

import java.net.URI;

// tag::download[]
public abstract class Download extends DefaultTask {
    @Input
    public abstract Property<String> getLocation();

    @Internal
    public Provider<URI> getUri() {
        return getLocation().map(l -> URI.create("https://" + l));
    }

    @TaskAction
    void run() {
        System.out.println("Downloading " + getUri().get());  // Use the `uri` provider (read-only property)
    }
}
// end::download[]

File: subprojects/docs/src/snippets/plugins/mutableManagedProperty/kotlin/buildSrc/src/main/kotlin/Download.kt
import org.gradle.api.DefaultTask
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.TaskAction
import java.net.URI

abstract class Download : DefaultTask() {
    // Use an abstract var
    @get:Input
    abstract var uri: URI

    @TaskAction
    fun run() {
        // Use the `uri` property
        println("Downloading $uri")
    }
}
